src/twisted/web/test/test_flatten.py

```diff
@@ -453,6 +457,53 @@ def test_unknownTypeRaises(self) -> None:
         self.assertFlatteningRaises(None, UnsupportedType)  # type: ignore[arg-type]
 
 
+class FlattenChunkingTests(SynchronousTestCase):
+    """
+    Tests for the way pieces of the result are chunked together in calls to
+    the write function.
+    """
+
+    def test_oneSmallChunk(self) -> None:
+        """
+        If the entire value to be flattened is available synchronously and fits
+        into the buffer it is all passed to a single call to the write
+        function.
+        """
+        output: List[bytes] = []
+        self.successResultOf(flatten(None, ["1", "2", "3"], output.append))
+        assert_that(output, equal_to([b"123"]))
+
+    def test_someLargeChunks(self) -> None:
+        """
+        If the entire value to be flattened is available synchronously but does
+        not fit into the buffer then it is chunked into buffer-sized pieces
+        and these are passed to the write function.
+        """
+        some = ["x"] * BUFFER_SIZE
+        someMore = ["y"] * BUFFER_SIZE
+        evenMore = ["z"] * BUFFER_SIZE
+
+        output: List[bytes] = []
+        self.successResultOf(flatten(None, [some, someMore, evenMore], output.append))
+        assert_that(
+            output,
+            equal_to([b"x" * BUFFER_SIZE, b"y" * BUFFER_SIZE, b"z" * BUFFER_SIZE]),
+        )
+
+    def test_chunksSeparatedByAsync(self) -> None:
+        """
+        When a L{Deferred} is encountered any buffered data is passed to the write
+        function.  After the result of the L{Deferred} is available it is
+        passed to another write along with following synchronous values.
+        """
+        c = succeed("c")
+        f = succeed("f")
+        value = ["a", "b", c, "d", "e", f, "g", "h"]
+        output: List[bytes] = []
+        self.successResultOf(flatten(None, value, output.append))
+        assert_that(output, equal_to([b"ab", b"cde", b"fgh"]))
```

Comment:
very minor suggestion

I start the review by checking the tests without checking the implementation and I was expecting it to block/wait only for deferred that don't have a result yet.

So this is why I was a bit confused by this test.

the `c` variable name blends in :)


```suggestion
        first_wait = succeed("first-")
        second_wait = succeed("second-")
        value = [
            "already-available",
            "-chunks",
            first_wait,
            "chunks-already-",
            "computed",
            second_wait,
            "more-chunks-",
            "already-available"
            ]
        output: List[bytes] = []
        self.successResultOf(flatten(None, value, output.append))
        assert_that(output, equal_to([
            b"already-available-chunks",
            b"first-chunks-already-computed",
            b"second-more-chunks-already-available"
            ]))
```


Maybe have a new test in which there is a check that the buffering is paused

```python
        first_wait = succeed("first-")
        second_wait = Deferred()
        value = [
            "already-available",
            "-chunks",
            first_wait,
            "chunks-already-",
            "computed",
            second_wait,
            "more-chunks-",
            "already-available"
            ]
        output: List[bytes] = []

        d = flatten(None, value, output.append)

        # Check d is not yet called.
        # and that we have a partial write.
        assert_that(output, equal_to([
            b"already-available-chunks",
            b"first-chunks-already-computed",
            ]))

        # We got a result and the flatten opration is now done.
        second_wait.callback('second-')
        self.successResultOf(d)
        assert_that(output, equal_to([
            b"already-available-chunks",
            b"first-chunks-already-computed",
            b"second-more-chunks-already-available"
            ]))
```